Modules/constants.c

```diff
@@ -6,373 +6,180 @@
+/* Convert an LDAP error into an informative python exception */
+PyObject*
+LDAPerror( LDAP *l, char *msg )
+{
+  if (l == NULL) {
+    PyErr_SetFromErrno( LDAPexception_class );
+    return NULL;
+  }
+  else {
+    int myerrno, errnum, opt_errnum;
+    PyObject *errobj;
+    PyObject *info;
+    PyObject *str;
+    PyObject *pyerrno;
+
+    /* at first save errno for later use before it gets overwritten by another call */
+    myerrno = errno;
+
+    char *matched, *error;
+
+    opt_errnum = ldap_get_option(l, LDAP_OPT_ERROR_NUMBER, &errnum);
+    if (opt_errnum != LDAP_OPT_SUCCESS)
+      errnum = opt_errnum;
+
+    if (errnum == LDAP_NO_MEMORY)
+      return PyErr_NoMemory();
+
+    if (errnum >= LDAP_ERROR_MIN && errnum <= LDAP_ERROR_MAX)
+      errobj = errobjects[errnum+LDAP_ERROR_OFFSET];
+    else
+      errobj = LDAPexception_class;
+
+    info = PyDict_New();
+    if (info == NULL)
+      return NULL;
+
+    str = PyUnicode_FromString(ldap_err2string(errnum));
+    if (str)
+      PyDict_SetItemString( info, "desc", str );
+    Py_XDECREF(str);
+
+    if (myerrno != 0) {
+        pyerrno = PyInt_FromLong(myerrno);
+        if (pyerrno)
+            PyDict_SetItemString( info, "errno", pyerrno );
+        Py_XDECREF(pyerrno);
+    }
 
-  obj = PyUnicode_FromString(LDAP_CONTROL_MANAGEDSAIT);
-  PyDict_SetItemString( d, "CONTROL_MANAGEDSAIT", obj );
-  Py_DECREF(obj);
+    if (ldap_get_option(l, LDAP_OPT_MATCHED_DN, &matched) >= 0
+      && matched != NULL) {
+        if (*matched != '\0') {
+      str = PyUnicode_FromString(matched);
+      if (str)
+          PyDict_SetItemString( info, "matched", str );
+      Py_XDECREF(str);
+        }
+        ldap_memfree(matched);
+    }
 
-  obj = PyUnicode_FromString(LDAP_CONTROL_PROXY_AUTHZ);
-  PyDict_SetItemString( d, "CONTROL_PROXY_AUTHZ", obj );
-  Py_DECREF(obj);
+    if (errnum == LDAP_REFERRAL) {
+        str = PyUnicode_FromString(msg);
+        if (str)
+      PyDict_SetItemString( info, "info", str );
+        Py_XDECREF(str);
+    } else if (ldap_get_option(l, LDAP_OPT_ERROR_STRING, &error) >= 0) {
+        if (error != NULL && *error != '\0') {
+            str = PyUnicode_FromString(error);
+            if (str)
+                PyDict_SetItemString( info, "info", str );
+            Py_XDECREF(str);
+        }
+        ldap_memfree(error);
+    }
+    PyErr_SetObject( errobj, info );
+    Py_DECREF(info);
+    return NULL;
+  }
+}
 
-  obj = PyUnicode_FromString(LDAP_CONTROL_SUBENTRIES);
-  PyDict_SetItemString( d, "CONTROL_SUBENTRIES", obj );
-  Py_DECREF(obj);
+/* initialise the module constants */
 
-  obj = PyUnicode_FromString(LDAP_CONTROL_VALUESRETURNFILTER);
-  PyDict_SetItemString( d, "CONTROL_VALUESRETURNFILTER", obj );
-  Py_DECREF(obj);
+int
+LDAPinit_constants( PyObject* m )
+{
+    PyObject *exc;
+    int result;
 
-  obj = PyUnicode_FromString(LDAP_CONTROL_ASSERT);
-  PyDict_SetItemString( d, "CONTROL_ASSERT", obj );
-  Py_DECREF(obj);
+#define check_result() {  \
```

Comment:
I don't like this kind of macros. It's confusing whenever a macro picks up values from the outer scope. Please either pass result in the function ``check(result);``or rewrite the few calls to ``if (PyModule_AddIntConstant(m, "OPT_ON", 1)) return -1;``

IIRC some compilers do not like a semicolon after a block ``{...};``. This could be a problem with MSVC.

Reply:
OK, I'll write out the error checking. (but I prefer an explicit `!= 0` if what I'm checking isn't a straightforward Boolean.)

Wrapped the macros in `do {...} while (0)`.